e2etest/clean/clean_ec2.go

```diff
@@ -0,0 +1,70 @@
+package main
+
+import (
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"log"
+	"time"
+)
+
+func main() {
+	// set up aws go sdk ec2 client
+	testSession, err := session.NewSession()
+	if err != nil {
+		log.Fatalf("Error creating session %v", err)
+	}
+	ec2client := ec2.New(testSession)
+
+	// Get list of instance
+	//State filter
+	instanceStateName := "instance-state-name"
+	instanceStateValue := "running"
+	instanceStateValues := []*string{&instanceStateValue}
+	instanceStateFilter := ec2.Filter{Name: &instanceStateName, Values: instanceStateValues}
+
+	//Tag filter
+	instanceTagName := "tag:Name"
+	instanceTagValueIntegTestSampleApp := "Integ-test-Sample-App"
+	instanceTagValueIntegTestAOC := "Integ-test-aoc"
+	instanceTagValues := []*string{&instanceTagValueIntegTestSampleApp, &instanceTagValueIntegTestAOC}
+	instanceTagFilter := ec2.Filter{Name: &instanceTagName, Values: instanceTagValues}
+
+	//get instances to delete
+	var deleteInstances []ec2.Instance
+	var nextToken *string
+	for {
+		describeInstancesInput := ec2.DescribeInstancesInput{Filters: []*ec2.Filter{&instanceStateFilter, &instanceTagFilter}, NextToken: nextToken}
+		describeInstancesOutput, err := ec2client.DescribeInstances(&describeInstancesInput)
+		if err != nil {
+			log.Fatalf("Failed to get instance for error %v", err)
+		}
+		for _, reservation := range describeInstancesOutput.Reservations {
+			for _, instance := range reservation.Instances {
+				//only delete instance if older than 30 days
+				if time.Now().UTC().Add(-1 * time.Hour * 24 * 30).After(*instance.LaunchTime) {
+					deleteInstances = append(deleteInstances, *instance)
+				}
+			}
+		}
+		if describeInstancesOutput.NextToken == nil {
+			break
+		}
+		nextToken = describeInstancesOutput.NextToken
+	}
+
+	//delete instances
+	var instanceIds []*string
+	for _, instance := range deleteInstances {
+		instanceIds = append(instanceIds, instance.InstanceId)
+		log.Printf("Delete instance %s tags %s launch-date %s", *instance.InstanceId, instance.Tags, instance.LaunchTime)
+	}
+	if len(instanceIds) < 1 {
+		log.Print("No instances to delete")
+		return
+	}
+	terminateInstancesInput := ec2.TerminateInstancesInput{InstanceIds: instanceIds}
```

Comment:
From the docs for [TerminateInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_TerminateInstances.html), the list of instance IDs can hold up to 1000 instances in a batch. Should we be concerned about hitting that limit in this cleanup job? If we're only ever spinning up a handful of instances, I'm more inclined to leave this as is, but if this is bound to grow, then might as well implement the batching now so it doesn't bite us later.

Reply:
I don't think we will have over 1000 instances to delete since we will be cleaning every week and we only clean instances over 1 month old.

Reply:
We had only 350 open instances and this has not been cleaned for almost 6 months.